mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala

```diff
@@ -0,0 +1,238 @@
+   *
+   * @param in input data
+   * @param isotonic asc or desc
+   * @return result
+   */
+  private def poolAdjacentViolators(
+      in: Array[(Double, Double, Double)],
+      isotonic: Boolean): Array[(Double, Double, Double)] = {
+
+    // Pools sub array within given bounds assigning weighted average value to all elements
+    def pool(in: Array[(Double, Double, Double)], start: Int, end: Int): Unit = {
+      val poolSubArray = in.slice(start, end + 1)
+
+      val weightedSum = poolSubArray.map(lp => lp._1 * lp._3).sum
+      val weight = poolSubArray.map(_._3).sum
+
+      var i = start
+      while (i <= end) {
+        in(i) = (weightedSum / weight, in(i)._2, in(i)._3)
+        i = i + 1
+      }
+    }
+
+    val monotonicityConstraintHolds: (Double, Double) => Boolean =
+      (x, y) => if (isotonic) x <= y else x >= y
+
+    var i = 0
+    while (i < in.length) {
+      var j = i
+
+      // Find monotonicity violating sequence, if any
+      while (j < in.length - 1 && !monotonicityConstraintHolds(in(j)._1, in(j + 1)._1)) {
+        j = j + 1
+      }
+
+      // If monotonicity was not violated, move to next data point
+      if (i == j) {
+        i = i + 1
+      } else {
+        // Otherwise pool the violating sequence
+        // And check if pooling caused monotonicity violation in previously processed points
+        while (i >= 0 && !monotonicityConstraintHolds(in(i)._1, in(i + 1)._1)) {
+          pool(in, i, j)
+          i = i - 1
+        }
+
+        i = j
+      }
+    }
+
+    in
+  }
+
+  /**
+   * Performs parallel pool adjacent violators algorithm
+   * Calls Pool adjacent violators on each partition and then again on the result
+   *
+   * @param testData input
+   * @param isotonic isotonic (increasing) or antitonic (decreasing) sequence
+   * @return result
+   */
+  private def parallelPoolAdjacentViolators(
+      testData: RDD[(Double, Double, Double)],
+      isotonic: Boolean): Array[(Double, Double, Double)] = {
+
+    val parallelStepResult = testData
+      .sortBy(_._2)
+      .mapPartitions(it => poolAdjacentViolators(it.toArray, isotonic).toIterator)
+
+    poolAdjacentViolators(parallelStepResult.collect(), isotonic)
+  }
+}
+
+/**
+ * Top-level methods for monotone regression (either isotonic or antitonic).
+ */
+object IsotonicRegression {
+
+  /**
+   * Train a monotone regression model given an RDD of (label, feature, weight).
+   * Label is the dependent y value
+   * Weight of the data point is the number of measurements. Default is 1
+   *
+   * @param input RDD of (label, feature, weight).
+   *              Each point describes a row of the data
+   *              matrix A as well as the corresponding right hand side label y
+   *              and weight as number of measurements
+   * @param isotonic isotonic (increasing) or antitonic (decreasing) sequence
+   */
+  def train(
+      input: RDD[(Double, Double, Double)],
+      isotonic: Boolean = true): IsotonicRegressionModel = {
```

Comment:
Same Java compatibility problem. We need to overload `train` to make it compatible with Java.

minor: I would like to remove the static train methods completely for easy maintenance.